airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java

```diff
@@ -190,32 +199,120 @@ PARTITION BY (DATE_TRUNC(_airbyte_extracted_at, DAY))
+            alterTableReport.columnsToChangeType(),
+            alterTableReport.isDestinationV2Format(),
+            tableClusteringMatches,
+            tablePartitioningMatches);
+
+    return alterTableReport.isNoOp() && tableClusteringMatches && tablePartitioningMatches;
+  }
+
+  @VisibleForTesting
+  public boolean clusteringMatches(StreamConfig stream, StandardTableDefinition existingTable) {
+    return existingTable.getClustering() == null ? false : containsAllIgnoreCase(
+            existingTable.getClustering().getFields().stream().collect(Collectors.toSet()),
+            clusteringColumns(stream));
+  }
+
+  @VisibleForTesting
+  public boolean partitioningMatches(StandardTableDefinition existingTable) {
+    return existingTable.getTimePartitioning() == null ? false : existingTable.getTimePartitioning()
+            .getField()
+            .equalsIgnoreCase("_airbyte_extracted_at") &&
+            TimePartitioning.Type.DAY.equals(existingTable.getTimePartitioning().getType());
+  }
+
+  @Override
+  public AlterTableReport buildAlterTableReport(final StreamConfig stream, final TableDefinition existingTable) {
+    final Map<String, StandardSQLTypeName> streamSchema = stream.columns().entrySet().stream()
+            .collect(Collectors.toMap(
+                    entry -> entry.getKey().name(),
+                    entry -> toDialectType(entry.getValue())
+            ));
+
+    final Map<String, StandardSQLTypeName> existingSchema = existingTable.getSchema().getFields().stream()
+            .collect(Collectors.toMap(
+                    field -> field.getName(),
+                    field -> field.getType().getStandardType()
+            ));
+
+    // Columns in the StreamConfig that don't exist in the TableDefinition
+    final Set<String> columnsToAdd = streamSchema.keySet().stream()
+            .filter(name -> !containsIgnoreCase(existingSchema.keySet(), name))
+            .collect(Collectors.toSet());
+
+    // Columns in the current schema that are no longer in the StreamConfig
+    final Set<String> columnsToRemove = existingSchema.keySet().stream()
+            .filter(name -> !containsIgnoreCase(streamSchema.keySet(), name) && !containsIgnoreCase(FINAL_TABLE_AIRBYTE_COLUMNS, name))
+            .collect(Collectors.toSet());
+
+    // Columns that are typed differently than the StreamConfig
+    final Set<String> columnsToChangeType = streamSchema.keySet().stream()
+            // If it's not in the existing schema, it should already be in the columnsToAdd Set
+            .filter(name -> {
+              // Big Query Columns are case-insensitive
+              final String matchedKey = matchingKey(existingSchema.keySet(), name).orElse(name);
+              return !existingSchema.getOrDefault(matchedKey, streamSchema.get(name)).equals(streamSchema.get(name));
+            })
+            .collect(Collectors.toSet());
+
+    final boolean isDestinationV2Format = FINAL_TABLE_AIRBYTE_COLUMNS.stream()
```

Comment:
do we have a testcase that exercises this?

Reply:
Added a unit test (and moved it to a different class), but should probably add an integration test